File: app/src/main/java/com/thatcakeid/zrytezene/ui/login/RegisterFragment.java
package com.thatcakeid.zrytezene.ui.login;

import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.android.material.button.MaterialButton;
import com.google.android.material.snackbar.Snackbar;
import com.google.android.material.textfield.TextInputEditText;
import com.google.android.material.textfield.TextInputLayout;
import com.google.firebase.FirebaseApp;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;
import com.thatcakeid.zrytezene.R;

public class RegisterFragment extends Fragment {
    private View view;
    private TextInputEditText tie1, tie2, tie3, tie4;
    private TextInputLayout til1, til2, til3, til4;
    private MaterialButton button_continue;
    private FirebaseAuth auth;

    public static RegisterFragment newInstance() {
        return new RegisterFragment();
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        view = inflater.inflate(R.layout.register_fragment, container, false);
        return view;
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        tie1 = view.findViewById(R.id.tie1);
        tie2 = view.findViewById(R.id.tie2);
        tie3 = view.findViewById(R.id.tie3);
        tie4 = view.findViewById(R.id.tie4);
        til1 = view.findViewById(R.id.til1);
        til2 = view.findViewById(R.id.til2);
        til3 = view.findViewById(R.id.til3);
        til4 = view.findViewById(R.id.til4);
        button_continue = view.findViewById(R.id.button_continue);

        FirebaseApp.initializeApp(getActivity());
        auth = FirebaseAuth.getInstance();

        button_continue.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (tie1.getText().toString().trim().length() == 0)
                    til1.setError("Email field can't be empty!");
                else {
                    if (tie2.getText().toString().trim().length() < 4)
                        til2.setError("Username field can't be empty or less than 4 characters!");
                    else {
                        if (tie3.getText().toString().length() == 0)
                            til3.setError("Password field can't be empty!");
                        else {
                            if (!tie3.getText().toString().equals(tie4.getText().toString()))
                                til4.setError("Password does not match");
                            else {
                                auth.createUserWithEmailAndPassword(tie1.getText().toString().trim(),
                                        tie3.getText().toString()).addOnSuccessListener(new OnSuccessListener<AuthResult>() {
                                    @Override
                                    public void onSuccess(AuthResult authResult) {
                                        auth.getCurrentUser().sendEmailVerification().addOnSuccessListener(new OnSuccessListener<Void>() {
                                            @Override
                                            public void onSuccess(Void aVoid) {
                                                Snackbar.make(view, "A verification link has been sent to your email. Please check your inbox or spam box.", Snackbar.LENGTH_LONG);
                                            }
                                        }).addOnFailureListener(new OnFailureListener() {
                                            @Override
                                            public void onFailure(@NonNull Exception e) {
                                                Snackbar.make(view, e.getMessage(), Snackbar.LENGTH_LONG);
                                            }
                                        });
                                        auth.signOut();
                                    }
                                }).addOnFailureListener(new OnFailureListener() {
                                    @Override
                                    public void onFailure(@NonNull Exception e) {
                                        Snackbar.make(view, e.getMessage(), Snackbar.LENGTH_LONG);
                                    }
                                });
                            }
                        }
                    }
                }
            }
        });
    }

}

File: app/src/main/java/com/thatcakeid/zrytezene/ui/startup/RegisterFragment.kt
package com.thatcakeid.zrytezene.ui.startup

import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import androidx.fragment.app.Fragment
import androidx.navigation.fragment.findNavController
import com.google.android.material.snackbar.Snackbar
import com.google.firebase.auth.FirebaseAuth
import com.thatcakeid.zrytezene.R
import com.thatcakeid.zrytezene.databinding.FragmentRegisterBinding
import com.zhuinden.fragmentviewbindingdelegatekt.viewBinding

// TODO: 10/5/21 move stuff to a viewmodel
class RegisterFragment : Fragment(R.layout.fragment_register) {
    private val binding: FragmentRegisterBinding by viewBinding(FragmentRegisterBinding::bind)
    private val auth by lazy { FirebaseAuth.getInstance() }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        binding.registerEmailTie.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            override fun afterTextChanged(s: Editable) {}

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                if (binding.registerEmailTie.text.toString().trim()
                        .matches("[a-zA-Z0-9._-]+@[a-z]+\\.+[a-z]+".toRegex())) {
                    binding.registerEmailTil.error = null
                } else {
                    binding.registerEmailTil.error = "Invalid email!"
                }
            }

        })

        requireArguments().getString("email")?.let {
            binding.registerEmailTie.setText(it)
        }

        binding.buttonContinue.setOnClickListener {

            val email = binding.registerEmailTie.text.toString().trim()
            val password = binding.registerPasswTie.text.toString().trim()
            val passwordRepeat = binding.registerPassw2Tie.text.toString().trim()

            if (!email.matches("[a-zA-Z0-9._-]+@[a-z]+\\.+[a-z]+".toRegex())) {
                Snackbar.make(binding.root, "Invalid email!", Snackbar.LENGTH_LONG).show()

                return@setOnClickListener
            }

            if (password.isEmpty()) {
                Snackbar.make(
                    binding.root,
                    "Password can't be empty!",
                    Snackbar.LENGTH_LONG
                ).show()

                return@setOnClickListener
            }

            if (password != passwordRepeat) {
                Snackbar.make(
                    binding.root,
                    "Password does not match",
                    Snackbar.LENGTH_LONG
                ).show()

                return@setOnClickListener
            }

            auth.createUserWithEmailAndPassword(
                binding.registerEmailTie.text.toString().trim(),
                binding.registerPasswTie.text.toString()
            ).addOnSuccessListener {
                assert(auth.currentUser != null)
                auth.currentUser!!
                    .sendEmailVerification().addOnSuccessListener {
                        Snackbar.make(
                            binding.root,
                            "A verification link has been sent to your email. Please check your inbox or spam box.",
                            Snackbar.LENGTH_LONG
                        ).show()
                    }
                    .addOnFailureListener { e ->
                        Snackbar.make(
                            binding.root,
                            e.message!!,
                            Snackbar.LENGTH_LONG
                        ).show()
                    }
                auth.signOut()
            }.addOnFailureListener { e ->
                Snackbar.make(
                    binding.root,
                    e.message!!,
                    Snackbar.LENGTH_LONG
                ).show()
            }
        }

        binding.textView5.setOnClickListener {
            findNavController()
                .popBackStack()
        }
    }
}